apimachinery/src/client.rs

```diff
@@ -0,0 +1,664 @@
+    resource: R,
+}
+
+impl<C, R> ResourceClient<C, R>
+where
+    C: ApiService + Send + Sync + Clone,
+    R: Resource,
+{
+    pub fn api_client(&self) -> &C {
+        &self.api_client
+    }
+}
+
+impl<C, R> ResourceClient<C, R>
+where
+    C: ApiService + Send + Sync + Clone,
+    R: Resource,
+{
+    pub fn get(
+        &self,
+        name: &R::Scope,
+        opts: GetOptions,
+    ) -> impl Future<Item = R::Item, Error = Error> + Send {
+        let gvr = self.resource.gvr();
+        let req = Request {
+            group: gvr.group.to_string(),
+            version: gvr.version.to_string(),
+            resource: gvr.resource.to_string(),
+            namespace: name.namespace().map(|s| s.to_string()),
+            name: name.name().map(|s| s.to_string()),
+            subresource: None,
+            method: Method::GET,
+            opts: opts,
+            content_type: None,
+            body: (),
+        };
+        self.api_client()
+            .request(req)
+            .map(|resp: Response<_>| resp.into_body())
+    }
+
+    pub fn create(
+        &self,
+        value: R::Item,
+        opts: GetOptions,
+    ) -> impl Future<Item = R::Item, Error = Error> + Send {
+        let ns = {
+            let metadata = value.metadata();
+            metadata.namespace.clone()
+        };
+        let gvr = self.resource.gvr();
+
+        let req = Request {
+            group: gvr.group.to_string(),
+            version: gvr.version.to_string(),
+            resource: gvr.resource.to_string(),
+            namespace: ns,
+            name: None,
+            subresource: None,
+            method: Method::POST,
+            opts: opts,
+            content_type: Some(APPLICATION_JSON),
+            body: value,
+        };
+        self.api_client()
+            .request(req)
+            .map(|resp: Response<_>| resp.into_body())
+    }
+
+    pub fn update(
+        &self,
+        value: R::Item,
+        opts: UpdateOptions,
+    ) -> impl Future<Item = R::Item, Error = Error> + Send {
+        let gvr = self.resource.gvr();
+        let (ns, name) = {
+            let md = value.metadata();
+            (md.namespace.clone(), md.name.clone())
+        };
+
+        let req = Request {
+            group: gvr.group.to_string(),
+            version: gvr.version.to_string(),
+            resource: gvr.resource.to_string(),
+            namespace: ns,
+            name: name,
+            subresource: None,
+            method: Method::PUT,
+            opts: opts,
+            content_type: Some(APPLICATION_JSON),
+            body: value,
+        };
+        self.api_client()
+            .request(req)
+            .map(|resp: Response<_>| resp.into_body())
+    }
+
+    pub fn patch(
+        &self,
+        name: R::Scope,
+        patch: Patch,
+        opts: UpdateOptions,
+    ) -> impl Future<Item = R::Item, Error = Error> + Send {
+        let gvr = self.resource.gvr();
+
+        let req = Request {
```

Comment:
The high repetition here makes me think a Request constructor from gvr + name + subresource + opts + contentype would help.

Reply:
Agreed 100%.  I was trying (unsuccessfully) to halt the rampant scope creep in this PR.  I will retro-fit a builder onto these in a followup.